### ISSUE TITLE
Add BatchNorm1D

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#78]
mean & variance should only be updated if the tape parameter actuall owns the tape. ie. `if TAPE::OWNS_TAPE`

The paper is: https://arxiv.org/abs/1502.03167

More details can be found at pytorch documentation: https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html

**Comment by vikigenius** [comment]
I was trying to implement this but am running into the following issue. We need to keep track of the running mean and running variance.

But the Module trait only allows for `fn forward(&self, input: Input) -> Self::Output;`

Shouldn't we need `fn forward(&mut self, input: Input) -> Self::Output;` if we want to update the running stats ? Or is there a way to avoid this that I am missing ?

**Comment by coreylowman** [comment]
Good call out. The absolute easiest option would be wrapping them in `RefCell` (like I did with the rng in Dropout/DropoutOneIn).  However now that there are at least 2 modules that would need a mut forward I'd want to think about how to actually support that.

Do you know if there are any other module types that would need `&mut self` in forward? I think a number of the norm layers require tracking running stats right? At least InstanceNorm does off the top of my head.

Changing all forwards to be `&mut` may be overkill from an api perspective since there aren't that many modules that require it, and I like how you can run forward on non-mutable models atm. But it'd be nice to make it clear when using batch norms that the module needs to be mut.

**Comment by vikigenius** [comment]
Yes, there is definitely InstanceNorm that would require it. But even apart from that if we ever get to the point where we start doing things like tracking Model metrics etc. we would need to allow forward to take a `&mut self` to update those metrics

**Comment by vikigenius** [comment]
Another possible decision to make is should we implement

```rust
impl<H: Tape, const M: usize> Module<Tensor1D<M, H>> for BatchNorm1D<M> 
```

Unlike layer norm batch norm is supposed to average across the batch dimension. When you have a 1D tensor is batch_size = 1? Or do you consider it as a batch of scalars. I am more inclined to treat it as a batch size of 1 and not implement it because the batch variance would always be zero in such  a case

**Comment by coreylowman** [comment]
Oh yeah this is gonna be subtle. Ran into something similar with Flatten (now called FlattenImage), cause for a 3d tensor it is most likely a single image rather than batches of 2d tensors. I am now seeing that pytorch BatchNorm1d only works for 2d & 3d tensors, and BatchNorm2d only works for 3d & 4d tensors, which addresses all of that.

So yeah I think it's okay to not implement for 1d or just return identity

**Comment by coreylowman** [comment]
Since BatchNorm2D was added, going to close this for now. This can be opened later if someone wants this, but this is very low priority for me
